### PR TITLE
Add app protect reconnect period directive

### PR DESCRIPTION
**Pull request by galitskiy**
### Proposed changes
Add _app_protect_reconnect_period_seconds_ directive, which determines the period of time between reconnect retries of the app-protect module to the web application firewall (WAF) engine.

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

File: internal/nginx/manager.go
Comment (ciarams87) [inline internal/nginx/manager.go:38]: What's the purpose of this change?

Comment (galitskiy) [reply internal/nginx/manager.go:38]: This is to give enough memory available for xml parser in order to avoid issuers with big policies. https://docs.nginx.com/nginx-app-protect/admin-guide/install/#:~:text=file.%20The%20number-,471859200,-should%20be%20enough

File: internal/configs/configmaps.go
Comment (lucacome) [inline internal/configs/configmaps.go:503]: Is a Float really needed? I would assume that seconds are int, float seems a little bit overkill 🤷‍♂️

File: internal/configs/configmaps.go
Comment (lucacome) [inline internal/configs/configmaps.go:508]: I find this a little bit confusing. If the value is between `0` and `60` I would assume that `0` and `60` are included. Is `60` a valid value?
Can you also add the valid values to the configmap key description?

**Review by lucacome — CHANGES_REQUESTED** [review]
Left a couple of comments, mostly about clarifying the allowed values and adding them to the docs.

File: internal/nginx/manager.go
Comment (lucacome) [reply internal/nginx/manager.go:38]: it's not really relevant to the reconnect period, is it?

File: internal/configs/configmaps.go
Comment (galitskiy) [reply internal/configs/configmaps.go:503]: Hi. according to docs "Value type: number with decimal fraction" (https://docs.nginx.com/nginx-app-protect/configuration-guide/configuration/)

File: internal/configs/configmaps.go
Comment (galitskiy) [reply internal/configs/configmaps.go:508]: Hi. the same here, its up to doc "Value Range: 0-60. 0 is illegal". so 60 is valid and 0 is not.

File: internal/nginx/manager.go
Comment (galitskiy) [reply internal/nginx/manager.go:37]: agree. create separate PR?

File: internal/configs/configmaps.go
Comment (lucacome) [reply internal/configs/configmaps.go:508]: I don't think it hurts to specify the values in the description and to make it more clear even if it doesn't match the one on the website.

File: internal/nginx/manager.go
Comment (lucacome) [reply internal/nginx/manager.go:37]: yeah I think it makese sense

Comment (galitskiy) [reply internal/nginx/manager.go]: created PR https://github.com/nginxinc/kubernetes-ingress/pull/2455

File: internal/configs/configmaps.go
Comment (galitskiy) [reply internal/configs/configmaps.go:508]: added some wording, please reopen if needed

**Review by ashutoshpradhan-f5 — APPROVED** [review]

File: docs/content/configuration/global-configuration/configmap-resource.md
Comment (tomasohaodha) [inline docs/content/configuration/global-configuration/configmap-resource.md:188]: I think this wording is strange, and may point to an underlying change that can make it more natural.
- "Decimal fraction" is not commonly used.
- "Range 0-60. 0 is illegal" is strange. Ranges are generally inclusive, or alternatively include the first number but exclude the second. I've never seen a range that excluded the first number.

Generally when time measurements require precision, I've seen milliseconds used, not seconds. This is the equivalent of seconds with three decimal places. It also makes the range description more readable, without the need for '0 is illegal'. We instead say "Range: 1-60000"

Can we either
a) Change unit to milliseconds to keep precision and eliminate awkward wording, or
b) Keep as seconds but remove the precision (value always rounded to the next whole second)

**Review by tomasohaodha — CHANGES_REQUESTED** [review]

File: docs/content/configuration/global-configuration/configmap-resource.md
Comment (jputrino) [reply docs/content/configuration/global-configuration/configmap-resource.md:188]: Agreed, this is unclear as written. However, as this is verbatim how the directive is described in the App Protect docs, this PR may not be the appropriate forum for requesting changes. It's sufficient to link to the directive in the NAP WAF docs here instead of duplicating the information.

```suggestion
|`app-protect-reconnect-period-seconds` | Sets the `app_protect_reconnect_period_seconds` [global directive](/nginx-app-protect/configuration/#global-directives). | `5` |  |
```

**Review by jputrino — APPROVED** [review]
Approved pending application of suggestion

File: docs/content/configuration/global-configuration/configmap-resource.md
Comment (tomasohaodha) [reply docs/content/configuration/global-configuration/configmap-resource.md:188]: Thanks @jputrino , that's a good way around it.

**Review by lucacome — APPROVED** [review]

**Review by tomasohaodha — APPROVED** [review]

File: internal/configs/configmaps.go
Comment (lucacome) [inline internal/configs/configmaps.go:504]: @galitskiy do we want to throw an error if the value is not between the accepted values? I think it would be easier to catch if we output something when the value is incorrect

File: internal/configs/configmaps.go
Comment (galitskiy) [inline internal/configs/configmaps.go:511]: @lucacome Thx for review. Here is suggestion to validate period. Please let me know if its' ugly or I misunderstood smth.
```suggestion
		if appProtectReconnectPeriod, exists, err := ParseFloat64(cfgm.Data["app-protect-reconnect-period-seconds"]); exists {
			if err != nil || appProtectReconnectPeriod <= 0 || appProtectReconnectPeriod > 60 {
				glog.Error("ConfigMap Key 'app-protect-reconnect-period-second' must have value between '0' and '60'. '0' is illegal. Ignoring.")
			} else {
				cfgParams.MainAppProtectReconnectPeriod = appProtectReconnectPeriod
			}
		}
	}
```

Comment (lucacome) [reply internal/configs/configmaps.go:511]: Unfortunately that doesn't work, I would do something lilke
```suggestion
		if appProtectReconnectPeriod, exists := cfgm.Data["app-protect-reconnect-period-seconds"]; exists {
			period, err := ParseFloat64(appProtectReconnectPeriod)
			if err == nil && period > 0 && period <= 60 {
				cfgParams.MainAppProtectReconnectPeriod = appProtectReconnectPeriod
			} else {
				glog.Error("ConfigMap Key 'app-protect-reconnect-period-second' must have value between '0' and '60'. '0' is illegal. Ignoring.")
			}
		}
	}
```

Comment (galitskiy) [reply internal/configs/configmaps.go:511]: yeah, you right. Thank you!

**Review by Jcahilltorre — APPROVED** [review]

**Review by lucacome — CHANGES_REQUESTED** [review]
The PR looks good, but I just noticed that the pipeline is not triggered. I'm going to block the merge until we sort that out.

**Review by lucacome — APPROVED** [review]

**Pull request merged**
